File: currency.js
/* ===== منتقي العملة وتحويل الأسعار المعروضة =====
   الأسعار محفوظة بالريال دائماً، وهذا الملفّ يحوّل المعروض
   وحده. الطلب والفاتورة يبقيان بالريال — وقاعدة البيانات
   تفرض ذلك بقيد check (currency = 'SAR') على الجدولين،
   وهو قيد مقصود لا نلتفّ عليه من المتصفّح.

   لماذا أسعار ثابتة لا خدمة حيّة: قوقل أغلقت واجهة أسعار
   الصرف سنة 2011، وما يُسمّى «أسعار قوقل» اليوم كشطٌ لصفحة
   البحث ينكسر بلا إنذار ويخالف شروطها. فالجدول هنا يعدّله
   المالك متى شاء، وبنية الملفّ تجعل وصل مزوّد حيّ لاحقاً
   تغييراً في دالّة واحدة (rate) لا في الصفحات.

   أهمّ ثلاثين عملة (بطلب المالك). الأساس: 1 ريال = كم من
   العملة. راجِعْها كلّ حين — الأسعار تتحرّك. */
(function (global) {
  "use strict";

  var STORE_KEY = "sf_currency";
  var BASE = "SAR";

  /* آخر مراجعة: 2026-09-08. القيمة = كم وحدة من هذه العملة
     يساوي ريالاً واحداً. الريال مربوط بالدولار عند 3.75،
     فسعره ثابت عمليّاً؛ وبقيّة العملات تتحرّك. */
  var TABLE = {
    SAR: { rate: 1,        ar: "ريال سعودي",      en: "Saudi Riyal",        sym: "﷼", dec: 2 },
    USD: { rate: 0.2667,   ar: "دولار أمريكي",    en: "US Dollar",          sym: "$",      dec: 2 },
    EUR: { rate: 0.2453,   ar: "يورو",            en: "Euro",               sym: "€",      dec: 2 },
    GBP: { rate: 0.2088,   ar: "جنيه إسترليني",   en: "British Pound",      sym: "£",      dec: 2 },
    AED: { rate: 0.9793,   ar: "درهم إماراتي",    en: "UAE Dirham",         sym: "د.إ",    dec: 2 },
    KWD: { rate: 0.0817,   ar: "دينار كويتي",     en: "Kuwaiti Dinar",      sym: "د.ك",    dec: 3 },
    QAR: { rate: 0.9709,   ar: "ريال قطري",       en: "Qatari Riyal",       sym: "ر.ق",    dec: 2 },
    BHD: { rate: 0.1005,   ar: "دينار بحريني",    en: "Bahraini Dinar",     sym: "د.ب",    dec: 3 },
    OMR: { rate: 0.1027,   ar: "ريال عُماني",     en: "Omani Rial",         sym: "ر.ع",    dec: 3 },
    EGP: { rate: 12.95,    ar: "جنيه مصري",       en: "Egyptian Pound",     sym: "ج.م",    dec: 2 },
    JOD: { rate: 0.1891,   ar: "دينار أردني",     en: "Jordanian Dinar",    sym: "د.أ",    dec: 3 },
    TRY: { rate: 10.92,    ar: "ليرة تركية",      en: "Turkish Lira",       sym: "₺",      dec: 2 },
    CNY: { rate: 1.897,    ar: "يوان صيني",       en: "Chinese Yuan",       sym: "¥",      dec: 2 },
    INR: { rate: 23.52,    ar: "روبية هندية",     en: "Indian Rupee",       sym: "₹",      dec: 2 },
    PKR: { rate: 74.85,    ar: "روبية باكستانية", en: "Pakistani Rupee",    sym: "₨",      dec: 0 },
    BDT: { rate: 32.28,    ar: "تاكا بنغلاديشي",  en: "Bangladeshi Taka",   sym: "৳",      dec: 0 },
    IDR: { rate: 4355,     ar: "روبية إندونيسية", en: "Indonesian Rupiah",  sym: "Rp",     dec: 0 },
    MYR: { rate: 1.124,    ar: "رنغيت ماليزي",    en: "Malaysian Ringgit",  sym: "RM",     dec: 2 },
    JPY: { rate: 39.45,    ar: "ين ياباني",       en: "Japanese Yen",       sym: "¥",      dec: 0 },
    KRW: { rate: 361.2,    ar: "وون كوري",        en: "South Korean Won",   sym: "₩",      dec: 0 },
    CHF: { rate: 0.2131,   ar: "فرنك سويسري",     en: "Swiss Franc",        sym: "Fr",     dec: 2 },
    CAD: { rate: 0.3648,   ar: "دولار كندي",      en: "Canadian Dollar",    sym: "C$",     dec: 2 },
    AUD: { rate: 0.4021,   ar: "دولار أسترالي",   en: "Australian Dollar",  sym: "A$",     dec: 2 },
    RUB: { rate: 21.34,    ar: "روبل روسي",       en: "Russian Ruble",      sym: "₽",      dec: 2 },
    ZAR: { rate: 4.712,    ar: "راند جنوب أفريقي", en: "South African Rand", sym: "R",     dec: 2 },
    NGN: { rate: 398.6,    ar: "نايرا نيجيري",    en: "Nigerian Naira",     sym: "₦",      dec: 0 },
    MAD: { rate: 2.451,    ar: "درهم مغربي",      en: "Moroccan Dirham",    sym: "د.م",    dec: 2 },
    TND: { rate: 0.8194,   ar: "دينار تونسي",     en: "Tunisian Dinar",     sym: "د.ت",    dec: 3 },
    SGD: { rate: 0.3417,   ar: "دولار سنغافوري",  en: "Singapore Dollar",   sym: "S$",     dec: 2 },
    BRL: { rate: 1.428,    ar: "ريال برازيلي",    en: "Brazilian Real",     sym: "R$",     dec: 2 }
  };

  var current = BASE;

  try {
    var saved = global.localStorage && localStorage.getItem(STORE_KEY);
    if (saved && TABLE[saved]) current = saved;
  } catch (e) { /* الوضع الخاصّ يمنع التخزين — يبقى الريال */ }

  function getCode() { return current; }

  function info(code) { return TABLE[code] || TABLE[BASE]; }

  function list() {
    var out = [];
    for (var k in TABLE) {
      if (Object.prototype.hasOwnProperty.call(TABLE, k)) {
        out.push({ code: k, ar: TABLE[k].ar, en: TABLE[k].en, sym: TABLE[k].sym });
      }
    }
    return out;
  }

  /* تحويل مبلغ بالريال إلى العملة الحالية. */
  function convert(sarValue) {
    var n = Number(sarValue);
    if (!isFinite(n)) return null;
    return n * info(current).rate;
  }

  /* تنسيق الرقم بفواصل الآلاف وبعدد الكسور المناسب لكلّ
     عملة: الين والوون بلا كسور، والدينار الكويتي بثلاثة. */
  function formatNumber(value, code) {
    var meta = info(code || current);
    var n = Number(value);
    if (!isFinite(n)) return "";
    try {
      return n.toLocaleString(undefined, {
        minimumFractionDigits: meta.dec,
        maximumFractionDigits: meta.dec
      });
    } catch (e) {
      return n.toFixed(meta.dec);
    }
  }

  function setCode(code) {
    if (!TABLE[code]) return false;
    current = code;
    try {
      if (global.localStorage) localStorage.setItem(STORE_KEY, code);
    } catch (e) { /* لا يمنع التغيير */ }
    return true;
  }

  /* هل المعروض عملة أجنبية؟ الصفحات تستعمله لتُظهر
     تنبيه «يُحاسَب بالريال» عند الدفع. */
  function isForeign() { return current !== BASE; }

  global.SFCurrency = {
    BASE: BASE,
    getCode: getCode,
    setCode: setCode,
    list: list,
    info: info,
    convert: convert,
    formatNumber: formatNumber,
    isForeign: isForeign
  };
})(window);

File: app_flutter/lib/core/currency.dart
import 'package:flutter/foundation.dart';
import 'package:intl/intl.dart';
import 'package:shared_preferences/shared_preferences.dart';

class SFCurrencyInfo {
  const SFCurrencyInfo(
    this.code,
    this.rate,
    this.ar,
    this.en,
    this.symbol,
    this.decimals,
  );
  final String code;
  final double rate;
  final String ar;
  final String en;
  final String symbol;
  final int decimals;
}

/// نفس أسعار العرض المرجعية في currency.js؛ الطلبات والفواتير تبقى بالريال.
class SFCurrency extends ChangeNotifier {
  SFCurrency._();
  static final instance = SFCurrency._();
  static const currencies = <SFCurrencyInfo>[
    SFCurrencyInfo('SAR', 1, 'ريال سعودي', 'Saudi Riyal', '﷼', 2),
    SFCurrencyInfo('USD', .2667, 'دولار أمريكي', 'US Dollar', '\$', 2),
    SFCurrencyInfo('EUR', .2453, 'يورو', 'Euro', '€', 2),
    SFCurrencyInfo('GBP', .2088, 'جنيه إسترليني', 'British Pound', '£', 2),
    SFCurrencyInfo('AED', .9793, 'درهم إماراتي', 'UAE Dirham', 'د.إ', 2),
    SFCurrencyInfo('KWD', .0817, 'دينار كويتي', 'Kuwaiti Dinar', 'د.ك', 3),
    SFCurrencyInfo('QAR', .9709, 'ريال قطري', 'Qatari Riyal', 'ر.ق', 2),
    SFCurrencyInfo('BHD', .1005, 'دينار بحريني', 'Bahraini Dinar', 'د.ب', 3),
    SFCurrencyInfo('OMR', .1027, 'ريال عُماني', 'Omani Rial', 'ر.ع', 3),
    SFCurrencyInfo('EGP', 12.95, 'جنيه مصري', 'Egyptian Pound', 'ج.م', 2),
    SFCurrencyInfo('JOD', .1891, 'دينار أردني', 'Jordanian Dinar', 'د.أ', 3),
    SFCurrencyInfo('TRY', 10.92, 'ليرة تركية', 'Turkish Lira', '₺', 2),
    SFCurrencyInfo('CNY', 1.897, 'يوان صيني', 'Chinese Yuan', '¥', 2),
    SFCurrencyInfo('INR', 23.52, 'روبية هندية', 'Indian Rupee', '₹', 2),
    SFCurrencyInfo('PKR', 74.85, 'روبية باكستانية', 'Pakistani Rupee', '₨', 0),
    SFCurrencyInfo('BDT', 32.28, 'تاكا بنغلاديشي', 'Bangladeshi Taka', '৳', 0),
    SFCurrencyInfo(
      'IDR',
      4355,
      'روبية إندونيسية',
      'Indonesian Rupiah',
      'Rp',
      0,
    ),
    SFCurrencyInfo('MYR', 1.124, 'رنغيت ماليزي', 'Malaysian Ringgit', 'RM', 2),
    SFCurrencyInfo('JPY', 39.45, 'ين ياباني', 'Japanese Yen', '¥', 0),
    SFCurrencyInfo('KRW', 361.2, 'وون كوري', 'South Korean Won', '₩', 0),
    SFCurrencyInfo('CHF', .2131, 'فرنك سويسري', 'Swiss Franc', 'Fr', 2),
    SFCurrencyInfo('CAD', .3648, 'دولار كندي', 'Canadian Dollar', 'C\$', 2),
    SFCurrencyInfo(
      'AUD',
      .4021,
      'دولار أسترالي',
      'Australian Dollar',
      'A\$',
      2,
    ),
    SFCurrencyInfo('RUB', 21.34, 'روبل روسي', 'Russian Ruble', '₽', 2),
    SFCurrencyInfo(
      'ZAR',
      4.712,
      'راند جنوب أفريقي',
      'South African Rand',
      'R',
      2,
    ),
    SFCurrencyInfo('NGN', 398.6, 'نايرا نيجيري', 'Nigerian Naira', '₦', 0),
    SFCurrencyInfo('MAD', 2.451, 'درهم مغربي', 'Moroccan Dirham', 'د.م', 2),
    SFCurrencyInfo('TND', .8194, 'دينار تونسي', 'Tunisian Dinar', 'د.ت', 3),
    SFCurrencyInfo(
      'SGD',
      .3417,
      'دولار سنغافوري',
      'Singapore Dollar',
      'S\$',
      2,
    ),
    SFCurrencyInfo('BRL', 1.428, 'ريال برازيلي', 'Brazilian Real', 'R\$', 2),
  ];

  SFCurrencyInfo _current = currencies.first;
  String get code => _current.code;
  SFCurrencyInfo get current => _current;
  double get rate => _current.rate;
  int get decimals => _current.decimals;
  bool get isForeign => code != 'SAR';

  Future<void> load() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final saved = prefs.getString('sf_currency');
      _current = currencies.firstWhere(
        (item) => item.code == saved,
        orElse: () => currencies.first,
      );
      notifyListeners();
    } catch (_) {
      // تعذّر التخزين لا يمنع عرض الأسعار بالريال.
    }
  }

  Future<void> setCurrency(String code) async {
    final matches = currencies.where((item) => item.code == code);
    if (matches.isEmpty || this.code == code) return;
    _current = matches.first;
    notifyListeners();
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString('sf_currency', code);
    } catch (_) {
      // الاختيار يظل فعّالاً للجلسة إذا تعذّر حفظه.
    }
  }

  double convert(double sar) => sar * rate;
  double roundConverted(double sar) =>
      double.parse(convert(sar).toStringAsFixed(decimals));

  /// مجموع العرض يساوي سعر الوحدة الظاهر مضروباً في الكمية، كما في الويب.
  double lineAmount(double sarUnit, int quantity) =>
      roundConverted(sarUnit) * quantity / rate;

  String format(double sar, {bool withSymbol = true}) {
    if (!sar.isFinite) return '—';
    final number = NumberFormat.decimalPatternDigits(
      locale: 'en',
      decimalDigits: decimals,
    ).format(convert(sar));
    return withSymbol ? '$number ${_current.symbol}' : number;
  }

  static String formatSar(double sar) =>
      '${NumberFormat.decimalPatternDigits(locale: 'en', decimalDigits: 2).format(sar)} ﷼';
}
